File: examples/tailcat-e2b/python/tailcat_e2b/tailcat_server.py
"""Start Tailcat listeners and wait until their addresses are reachable."""

import os
import re
import shutil
import subprocess
import tempfile
import time
from dataclasses import dataclass
from typing import Callable

from e2b import Sandbox

from .e2b_sandbox import CommandRunner, run_sandbox_command

TAILCAT_ADDRESS_PATTERN = re.compile(r"\btc[A-Za-z0-9_-]{20,}")


@dataclass
class TailcatServer:
    address: str
    stop: Callable[[], None]


@dataclass
class SandboxTailcatServer(TailcatServer):
    # Wait for a one-shot listener (bare `tailcat`) to exit and return what it wrote to stdout.
    wait: Callable[[], str]


def start_sandbox_tailcat_server(
    sandbox: Sandbox,
    tailcat_arguments: str,
    instance_name: str = "tailcat",
    wait_seconds: float = 20,
) -> SandboxTailcatServer:
    """Start a Tailcat listener inside a sandbox and read its generated address."""
    address_file = f"/tmp/{instance_name}.addr"
    log_file = f"/tmp/{instance_name}.log"
    # exec: the SDK kills the wrapping shell by pid, so tailcat must be that process.
    server_process = sandbox.commands.run(
        f"rm -f {address_file}; TAILCAT_ADDR_FILE={address_file} exec tailcat {tailcat_arguments} 2> {log_file}",
        background=True,
    )

    def read_address_file() -> str:
        return run_sandbox_command(sandbox, f"cat {address_file} 2>/dev/null").output

    def read_server_log() -> str:
        return run_sandbox_command(sandbox, f"cat {log_file}").output

    try:
        address = _wait_for_tailcat_address(read_address_file, wait_seconds, read_server_log)
    except Exception:
        server_process.kill()
        raise
    print(f"[{sandbox.sandbox_id}] tailcat {tailcat_arguments}  ->  {_abbreviate(address)}")
    return SandboxTailcatServer(address, stop=server_process.kill, wait=lambda: server_process.wait().stdout)


def start_local_tailcat_server(tailcat_arguments: str, wait_seconds: float = 20) -> TailcatServer:
    """Start a Tailcat listener on the laptop and read its generated address."""
    require_local_tailcat()
    address_file = os.path.join(tempfile.gettempdir(), f"tailcat-{os.getpid()}-{int(time.time() * 1000)}.addr")
    server_process = subprocess.Popen(
        ["tailcat", *tailcat_arguments.split()],
        env={**os.environ, "TAILCAT_ADDR_FILE": address_file},
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )

    def read_address_file() -> str:
        if not os.path.exists(address_file):
            return ""
        with open(address_file) as file:
            return file.read()

    try:
        address = _wait_for_tailcat_address(read_address_file, wait_seconds)
    except Exception:
        server_process.terminate()
        raise

    print(f"[laptop] tailcat {tailcat_arguments}  ->  {_abbreviate(address)}")

    def stop() -> None:
        server_process.terminate()
        if os.path.exists(address_file):
            os.unlink(address_file)

    return TailcatServer(address, stop)


def wait_until_reachable(run_command: CommandRunner, address: str, timeout_seconds: float = 60) -> str:
    """Retry Tailcat ping until the listener has joined DERP, then report its path.

    Each attempt lets tailcat resend its rendezvous ping for up to 10 s; the loop
    covers a relay that admits the listener later than that.
    """
    started_at = time.time()
    last_output = ""
    while time.time() - started_at < timeout_seconds:
        ping = run_command(f"tailcat ping --timeout 10s {address}")
        if ping.exit_code == 0:
            pong_line = next((line for line in ping.output.splitlines() if "pong in" in line), None)
            if pong_line and "DERP" in pong_line:
                return f"DERP relay: {pong_line.split('pong in ', maxsplit=1)[-1]}"
            return f"Tailcat connection: {pong_line.strip() if pong_line else 'ready'}"
        last_output = ping.output
        time.sleep(1)
    raise RuntimeError(
        f"tailcat server {_abbreviate(address)} not reachable after {timeout_seconds}s:\n{last_output}"
    )


def require_local_tailcat() -> None:
    if not shutil.which("tailcat"):
        raise RuntimeError(
            "tailcat is not installed; run `brew install tailcat` or see "
            "https://github.com/tailscale/tailcat#install"
        )


def _wait_for_tailcat_address(
    read_address_file: Callable[[], str],
    wait_seconds: float,
    read_server_log: Callable[[], str] | None = None,
) -> str:
    started_at = time.time()
    while time.time() - started_at < wait_seconds:
        if address_match := TAILCAT_ADDRESS_PATTERN.search(read_address_file()):
            return address_match.group(0)
        time.sleep(0.3)
    server_log = f"\n{read_server_log()}" if read_server_log else ""
    raise RuntimeError(f"tailcat did not print an address within {wait_seconds}s{server_log}")


def _abbreviate(address: str) -> str:
    return f"{address[:20]}…"

File: examples/tailcat-e2b/js/src/tailcatServer.ts
import { spawn, spawnSync, type ChildProcess } from "node:child_process";
import { existsSync, readFileSync, unlinkSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { setTimeout as delay } from "node:timers/promises";
import type { CommandHandle, Sandbox } from "e2b";
import { type CommandRunner, runSandboxCommand } from "./e2bSandbox";

const TAILCAT_ADDRESS_PATTERN = /\btc[A-Za-z0-9_-]{20,}/;

export interface TailcatServer {
  address: string;
  stop(): Promise<void> | void;
}

export interface SandboxTailcatServer extends TailcatServer {
  /** Wait for a one-shot listener (bare `tailcat`) to exit and return what it wrote to stdout. */
  wait(): Promise<string>;
}

/** Start a Tailcat listener inside a sandbox and read its generated address. */
export async function startSandboxTailcatServer(
  sandbox: Sandbox,
  tailcatArguments: string,
  instanceName = "tailcat",
  waitMs = 20_000,
): Promise<SandboxTailcatServer> {
  const addressFile = `/tmp/${instanceName}.addr`;
  const logFile = `/tmp/${instanceName}.log`;
  const serverProcess: CommandHandle = await sandbox.commands.run(
    // exec: the SDK kills the wrapping shell by pid, so tailcat must be that process.
    `rm -f ${addressFile}; TAILCAT_ADDR_FILE=${addressFile} exec tailcat ${tailcatArguments} 2> ${logFile}`,
    { background: true },
  );

  async function readAddressFile(): Promise<string> {
    return (await runSandboxCommand(sandbox, `cat ${addressFile} 2>/dev/null`))
      .output;
  }

  async function readServerLog(): Promise<string> {
    return (await runSandboxCommand(sandbox, `cat ${logFile}`)).output;
  }

  let address: string;
  try {
    address = await waitForTailcatAddress(
      readAddressFile,
      waitMs,
      readServerLog,
    );
  } catch (error) {
    await serverProcess.kill();
    throw error;
  }
  console.log(
    `[${sandbox.sandboxId}] tailcat ${tailcatArguments}  ->  ${abbreviate(
      address,
    )}`,
  );

  return {
    address,
    async stop(): Promise<void> {
      await serverProcess.kill();
    },
    async wait(): Promise<string> {
      return (await serverProcess.wait()).stdout;
    },
  };
}

/** Start a Tailcat listener on the laptop and read its generated address. */
export async function startLocalTailcatServer(
  tailcatArguments: string,
  waitMs = 20_000,
): Promise<TailcatServer> {
  requireLocalTailcat();
  const addressFile = join(
    tmpdir(),
    `tailcat-${process.pid}-${Date.now()}.addr`,
  );
  const serverProcess: ChildProcess = spawn(
    "tailcat",
    tailcatArguments.split(/\s+/).filter(Boolean),
    {
      env: { ...process.env, TAILCAT_ADDR_FILE: addressFile },
      stdio: "ignore",
    },
  );

  function readAddressFile(): Promise<string> {
    return Promise.resolve(
      existsSync(addressFile) ? readFileSync(addressFile, "utf8") : "",
    );
  }

  let address: string;
  try {
    address = await waitForTailcatAddress(readAddressFile, waitMs);
  } catch (error) {
    serverProcess.kill();
    throw error;
  }
  console.log(
    `[laptop] tailcat ${tailcatArguments}  ->  ${abbreviate(address)}`,
  );

  return {
    address,
    stop(): void {
      serverProcess.kill();
      if (existsSync(addressFile)) {
        unlinkSync(addressFile);
      }
    },
  };
}

/**
 * Retry Tailcat ping until the listener has joined DERP, then report its path.
 * Each attempt lets tailcat resend its rendezvous ping for up to 10 s; the loop
 * covers a relay that admits the listener later than that.
 */
export async function waitUntilReachable(
  runCommand: CommandRunner,
  address: string,
  timeoutMs = 60_000,
): Promise<string> {
  const startedAt = Date.now();
  let lastOutput = "";
  while (Date.now() - startedAt < timeoutMs) {
    const ping = await runCommand(`tailcat ping --timeout 10s ${address}`);
    if (ping.exitCode === 0) {
      const pongLine = ping.output
        .split("\n")
        .find((line) => line.includes("pong in"));
      return pongLine?.includes("DERP")
        ? `DERP relay: ${pongLine.split("pong in ")[1] ?? pongLine.trim()}`
        : `Tailcat connection: ${pongLine?.trim() ?? "ready"}`;
    }
    lastOutput = ping.output;
    await delay(1000);
  }
  throw new Error(
    `tailcat server ${abbreviate(address)} not reachable after ${
      timeoutMs / 1000
    }s:\n${lastOutput}`,
  );
}

export function requireLocalTailcat(): void {
  if (spawnSync("tailcat", ["version"], { stdio: "ignore" }).status !== 0) {
    throw new Error(
      "tailcat is not installed; run `brew install tailcat` or see https://github.com/tailscale/tailcat#install",
    );
  }
}

async function waitForTailcatAddress(
  readAddressFile: () => Promise<string>,
  waitMs: number,
  readServerLog?: () => Promise<string>,
): Promise<string> {
  const startedAt = Date.now();
  while (Date.now() - startedAt < waitMs) {
    const addressMatch = TAILCAT_ADDRESS_PATTERN.exec(await readAddressFile());
    if (addressMatch) {
      return addressMatch[0];
    }
    await delay(300);
  }
  const serverLog = readServerLog ? `\n${await readServerLog()}` : "";
  throw new Error(
    `tailcat did not print an address within ${waitMs / 1000}s${serverLog}`,
  );
}

function abbreviate(address: string): string {
  return `${address.slice(0, 20)}…`;
}
